contrib/babelfishpg_tsql/src/rolecmds.c

```diff
@@ -712,42 +712,69 @@ PG_FUNCTION_INFO_V1(user_id);
 Datum
 user_id(PG_FUNCTION_ARGS)
 {
-	char	   *user_input;
-	char	   *user_name;
-	char	   *db_name;
-	HeapTuple	auth_tuple;
-	Form_pg_authid authform;
-	Oid			ret;
+    char    *user_input;
+    char    *user_name;
+    char    *db_name;
+    HeapTuple    auth_tuple;
+    Form_pg_authid   authform;
+    Oid     ret;
+    size_t  len;
 
-	user_input = PG_ARGISNULL(0) ? NULL : text_to_cstring(PG_GETARG_TEXT_PP(0));
-	db_name = get_cur_db_name();
+    user_input = PG_ARGISNULL(0) ? NULL : text_to_cstring(PG_GETARG_TEXT_PP(0));
+    db_name = get_cur_db_name();
 
-	if (!db_name)
-		PG_RETURN_NULL();
+    if (!db_name)
+        PG_RETURN_NULL();
 
-	if (!user_input)
-		PG_RETURN_OID(GetUserId());
+    if (!user_input)
+        PG_RETURN_OID(GetUserId());
 
-	user_name = get_physical_user_name(db_name, user_input);
+    user_name = get_physical_user_name(db_name, user_input);
 
-	if (!user_name)
-		PG_RETURN_NULL();
+    if (!user_name)
+        PG_RETURN_NULL();
 
-	if (pltsql_case_insensitive_identifiers)
-		/* Lowercase the entry, if needed */
-		for (char *p = user_name; *p; ++p)
-			*p = tolower(*p);
+    len = strlen(user_name);
+    while (len > 0 && isspace(user_name[len-1]))
+        user_name[--len] = '\0';
 
-	auth_tuple = SearchSysCache1(AUTHNAME, CStringGetDatum(user_name));
-	if (!HeapTupleIsValid(auth_tuple))
-		PG_RETURN_NULL();
+    if (pltsql_case_insensitive_identifiers)
+    {
+        /* Lowercase the entry, if needed */
+        for (char *p = user_name; *p; ++p)
+            *p = tolower(*p);
+    }
 
-	authform = (Form_pg_authid) GETSTRUCT(auth_tuple);
-	ret = authform->oid;
+    auth_tuple = SearchSysCache1(AUTHNAME, CStringGetDatum(user_name));
 
-	ReleaseSysCache(auth_tuple);
+    if (!HeapTupleIsValid(auth_tuple))
+        PG_RETURN_NULL();
+    authform = (Form_pg_authid) GETSTRUCT(auth_tuple);
+    ret = authform->oid;
 
-	PG_RETURN_OID(ret);
+    ReleaseSysCache(auth_tuple);
+
+    PG_RETURN_OID(ret);
+}
+
+PG_FUNCTION_INFO_V1(user_id_noarg);
+Datum
+user_id_noarg(PG_FUNCTION_ARGS)
+{
+    HeapTuple auth_tuple;
+    Form_pg_authid authform;
+    Oid ret;
+
+    auth_tuple = SearchSysCache1(AUTHNAME, CStringGetDatum(GetUserNameFromId(GetUserId(), false)));
```

Comment:
GetUserId() = authform->oid in your case.  
No need to search again, just return it. 

refer to line 730.